### ISSUE TITLE
[Question] SHMInitialize, SHMAlloc

### ISSUE THREAD

**Issue by ruben-ayrapetyan** [dotnet/coreclr#12765]
Hello.

There is following code in `shmemory.cpp`:
```
BOOL SHMInitialize(void)
{
    InternalInitializeCriticalSection(&shm_critsec);

    init_waste();
    
        int size;
        SHM_FIRST_HEADER *header;
        SHMPTR pool_start;
        SHMPTR pool_end;
        enum SHM_POOL_SIZES sps;

        TRACE("Now initializing global shared memory system\n");
        
        // Not really shared in CoreCLR; we don't try to talk to other CoreCLRs.
        shm_segment_bases[0] = mmap(NULL, segment_size,PROT_READ|PROT_WRITE,
                                    MAP_ANON|MAP_PRIVATE, -1, 0);
        if(shm_segment_bases[0] == MAP_FAILED)
        {
            ERROR("mmap() failed; error is %d (%s)\n", errno, strerror(errno));
            return FALSE;
        }
        TRACE("Mapped first SHM segment at %p\n",shm_segment_bases[0].Load());

  ...
```

Is the "`// Not really shared in CoreCLR; we don't try to talk to other CoreCLRs`" comment still correct? 
Then, the `SHM`-allocated memory is not shared with other processes in any case, is it so?

cc @lemmaa @Dmitri-Botcharnikov @gbalykov @kvochko

**Comment by ruben-ayrapetyan** [comment]
@janvorli, could you, please, take a look?

**Comment by ruben-ayrapetyan** [comment]
Could you, please, also clarify how the mechanism of file locking (`FILEGetSHMFileLocks`) does work?
What is the scope of locking?

**Comment by ruben-ayrapetyan** [comment]
As far as I currently undestand, the file locking emulates `FILE_SHARE_READ` / `FILE_SHARE_WRITE` semantics of Windows API `CreateFile` in scope of one CoreCLR instance. Is it correct?

**Comment by jkotas** [comment]
It used to implement shared state accross multiple processes. The code that did implement the shared state accross multiple processes was removed in https://github.com/dotnet/coreclr/commit/3433abc7838db54b9656bd0cfb606a7a0d6fed91. 

It does not have very meaningful purpose today - the left-over part is really just an over-engineered custom memory allocator. It should be fine to remove it, or simplify a lot at least.

**Comment by ruben-ayrapetyan** [comment]
@jkotas, Thank you.

**Comment by janvorli** [comment]
Fixed by #12815
